karoo_gp/base_class.py

```diff
@@ -14,182 +14,165 @@
+    │   ├─ .trees                       - list of Trees, length tree_pop_max
+    │   │   ├─ ...
+    │   │   └─ Tree                     - an evolvable expression tree
+    │   │      ├─ .expression           - a string of sympified expr, e.g. a*b
+    │   │      ├─ .score                - a dict of results matching `scoring`
+    │   │      ├─ .root = Branch        - a recursive node which forms a Tree
+    │   │      │   ├─ .symbol           - a terminal ('a') or function ('*')
+    │   │      │   ├─ .arity            - instructions for generating children
+    │   │      │   ├─ .parent           - the node immediately above
+    │   │      │   └─ .children         - the nodes immediately below
+    │   │      │       └─ ...
+    │   │      └─ mutate, crossover...  - methods used by population.evolve
+    │   │
+    │   ├─ .evaluate(X, y)              - predict and score trees
+    │   └─ .evolve()                    - return a new generation of trees
+    │
+    ├─ .fit(X, y)                       - evolve expressions to predict y
+    ├- .predict(X)                      - return predicted y values for X
+    └- .score(pred, y)                  - return score of prediction against y
+    """
+
+    def __init__(
+        self, tree_type='r', tree_depth_base=3, tree_depth_max=3,
+        tree_depth_min=1, tree_pop_max=100, gen_max=10, tourn_size=7,
+        filename='', output_dir='', evolve_repro=0.1, evolve_point=0.1,
+        evolve_branch=0.2, evolve_cross=0.6, display='s', precision=None,
+        swim='p', mode='s', seed=None, pause_callback=None,
+        engine_type='numpy', tf_device="/gpu:0", tf_device_log=False,
+        functions=None, terminals=None, test_size=0.2, scoring=None,
+        higher_is_better=False, cache={}):
+        """Initialize a Karoo_GP object with given parameters"""
+
+        # Model parameters
         self.seed = seed
-        self.rng = np.random.default_rng(seed)  # this is used by Karoo
-        # the following two are not used, set them just in case
-        np.random.seed(seed)  # this was used by sklearn while classifying
-        tf.set_random_seed(seed)
-
-        ### PART 2 - construct first generation of Trees ###
-        self.fx_data_load(filename)
-        self.functions = Functions([f[0] for f in self.functions])  # Symbol only
-        self.terminals = Terminals(self.terminals[:-1])
-
-        self.fx_fitness_labels_map = fx_fitness_labels_map_maker(self.class_labels)
-
-        self.log(f'\n\t\033[32m Press \033[36m\033[1m?\033[0;0m\033[32m at any '
-                 f'\033[36m\033[1m(pause)\033[0;0m\033[32m, or '
-                 f'\033[36m\033[1mENTER\033[0;0m \033[32mto continue the run\033[0;0m',
-                 display=['i'])
-        self.pause(display=['i'])
-
-        # initialise population_a to host the first generation
+        self.rng = np.random.default_rng(seed)
+        np.random.seed(seed)                 # used by skm in train_test_split
```

Comment:
FTR, this was changed in #60.

By reading your link, it's not clear to me that this needs to be changed.  The part you quoted only says that the function needs to accept a `random_state` arg, but doesn't say explicitly that it has to be a `RandomState` instance.

For example, `train_test_split` accepts a [`random_state` arg](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html), but the value can be either an `int` or a `RandomState` instance:
> `random_state`: ***int, RandomState instance or None, default=None***
>
>    Controls the shuffling applied to the data before applying the split. Pass an int for reproducible output across multiple function calls. See [Glossary](https://scikit-learn.org/stable/glossary.html#term-random_state).

The [`check_random_state`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html#sklearn.utils.check_random_state) mentioned in your quote also accepts either an `int` or a `RandomState` instance, so the estimator should be able to safely store either and pass them as they are to `check_random_state`.

Reply:
Right - it can also be an integer or None, but it doesn't accept a [`Generator`](https://numpy.org/doc/stable/reference/random/generator.html), which is what we're using now.

Looks like we could:
A) If an int is passed, use it to create a `RandomState`. If RandomState is passed, use that. for rng.
B) If an int is passed, use it to create the `Generator`. If RandomState is passed, store it but ignore and create/use a `Generator` instead (like we are now) for rng.